docs/build-as-dll.md

```diff
@@ -0,0 +1,8 @@
+# Design of building OpenTelemetry C++ SDK and exporters as DLL
+
+OpenTelemetry C++ can be built into single DLL named opentelemetry_cpp.dll on
+Windows, with CMake flag `OPENTELEMETRY_BUILD_DLL` set to `ON`.
+
+To instrument a application based on the DLL, please define macro
+`OPENTELEMETRY_BUILD_IMPORT_DLL` before including any OpenTelemetry API header
+files in the application source files.
```

Comment:
I am convinced this part can be simplified.

Once OPENTELEMETRY_BUILD_DDL is defined, the build uses a DDL.

The user code should not have to define OPENTELEMETRY_BUILD_IMPORT_DDL.

Instead, the makefiles in opentelemetry-cpp should know when a file that belong to the DDL is compiled (EXPORT) or when a file that does not belong to the DDL is compiled (IMPORT).

Reply:
The user mode can also built with the exiting static library which is default, so we need a new macro to guard referencing the DLL, so we don't break any existing users' instrumentation code.
> The user code should not have to define OPENTELEMETRY_BUILD_IMPORT_DLL.

Reply:
Trying one more time with code, to make my suggestion more clear.

The choice between import and export is driven by which file is compiled, not by the user code.

In CMakeList.txt:

```
if(DEFINED OPENTELEMETRY_BUILD_DLL)
  if(NOT MSVC)
    message(WARNING "Build DLL is supposed to work with MSVC!")
  endif()
  if(WITH_STL)
    message(
      WARNING "Build DLL with C++ STL could cause binary incompatibility!")
  endif()
  add_definitions(-DOPENTELEMETRY_BUILD_DLL) // <-------------------------
endif()
```

In macro.h:

```
#if defined(_MSC_VER) && defined(OPENTELEMETRY_BUILD_DLL)

#ifdef OPENTELEMETRY_BUILD_EXPORT_DLL // <-------------------------------------
#  define OPENTELEMETRY_EXPORT __declspec(dllexport)
#else
#  define OPENTELEMETRY_EXPORT __declspec(dllimport)
#endif

#else

//
// build OpenTelemetry as static library or not on Windows.
//
#  define OPENTELEMETRY_EXPORT

#endif
```

And then, when compiling every file that is part of the DLL:

File `sdk/src/trace/CMakeLists.txt`

```
add_definitions(-DOPENTELEMETRY_BUILD_EXPORT_DLL) // <------------------------------
```

This totally avoids user to add `OPENTELEMETRY_BUILD_IMPORT_DLL`.